pkg/reconciler/eventtype.go

```diff
@@ -0,0 +1,193 @@
+
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		LabelSelector: labels.SelectorFromSet(lbs),
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		el := &eventingv1alpha1.EventTypeList{}
+		if err := r.Client.List(ctx, opts, el); err != nil {
+			return nil, err
+		}
+
+		for _, e := range el.Items {
+			if metav1.IsControlledBy(&e, owner) {
+				eventTypes = append(eventTypes, e)
+			}
+		}
+		if el.Continue != "" {
+			opts.Raw.Continue = el.Continue
+		} else {
+			return eventTypes, nil
+		}
+	}
+}
+
+// makeEventTypes creates the in-memory representation of the EventTypes.
+func (r *EventTypeReconciler) makeEventTypes(args *EventTypeReconcilerArgs, owner metav1.Object) ([]eventingv1alpha1.EventType, error) {
+	eventTypes := make([]eventingv1alpha1.EventType, 0)
+	for _, spec := range args.EventTypeSpecs {
+		eventType := r.makeEventType(spec, args.Namespace, args.Labels)
+		// Setting the reference to delete the EventType upon uninstalling the source.
+		if err := controllerutil.SetControllerReference(owner, &eventType, r.Scheme); err != nil {
+			return nil, err
+		}
+		eventTypes = append(eventTypes, eventType)
+	}
+	return eventTypes, nil
+}
+
+// makeEventType creates the in-memory representation of the EventType.
+func (r *EventTypeReconciler) makeEventType(spec eventingv1alpha1.EventTypeSpec, namespace string, lbl map[string]string) eventingv1alpha1.EventType {
+	return eventingv1alpha1.EventType{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", ToDNS1123Subdomain(spec.Type)),
+			Labels:       lbl,
+			Namespace:    namespace,
+		},
+		Spec: spec,
+	}
+}
+
+// getSyncEventTypes computes the EventTypes that need to be created and/or deleted based on the difference between
+// 'expected' and 'current'. It does so using all the EventType.Spec fields but Description.
+func (r *EventTypeReconciler) getEventTypesToSync(current []eventingv1alpha1.EventType, expected []eventingv1alpha1.EventType) *syncEventTypes {
+	eventTypesToSync := &syncEventTypes{
+		ToCreate: make([]eventingv1alpha1.EventType, 0),
+		ToDelete: make([]eventingv1alpha1.EventType, 0),
+	}
+	keyFunc := func(eventType *eventingv1alpha1.EventType) string {
+		return fmt.Sprintf("%s_%s_%s_%s", eventType.Spec.Type, eventType.Spec.Source, eventType.Spec.Schema, eventType.Spec.Broker)
+	}
+
+	currentMap := groupByKey(current, keyFunc)
+	for _, e := range expected {
+		if et, ok := currentMap[keyFunc(&e)]; !ok {
+			// If it's not in the currentMap, we need to create it.
+			eventTypesToSync.ToCreate = append(eventTypesToSync.ToCreate, e)
+		} else {
+			// If it's in the currentMap and there are more than one,
+			// then we just keep the first one, and delete the others.
+			if len(et) > 1 {
+				for i := 1; i < len(et); i++ {
+					eventTypesToSync.ToDelete = append(eventTypesToSync.ToDelete, et[i])
```

Comment:
I wonder if there are race conditions here. I haven't thought it through carefully, but if two controllers are trying to delete event types, is it possible they could choose to keep different ones, thus deleting all of them? This list only includes types that are owned by a single source object, right?

I guess if we were worried about this, one possible workaround is to sort by creation timestamp (plus some deterministic secondary key like UID to break ties) so that the oldest one is always kept. Seems like a lot of work though.

@nachocano if the "clean up duplicates" task were done by a different controller whose only job was cleaning up EventType dupes, would this function be simpler (or be necessary at all)?

Reply:
I think this deletion logic is not needed anymore. As of now, EventType spec fields that determine their uniqueness (type,source,schema,broker) are immutable. The only thing that could happen is if someone deletes them by mistake, if so, then we recreate them. 
Will remove this deletion logic and simplify this function.

> This list only includes types that are owned by a single source object, right?

Yes, that is correct.